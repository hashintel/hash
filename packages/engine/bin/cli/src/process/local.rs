use async_trait::async_trait;
use error::{Report, Result, ResultExt};
use hash_engine::{nano, proto::EngineMsg};

use super::process;

#[cfg(debug_assertions)]
const PROCESS_PATH_DEFAULT: &str = "./target/debug/hash_engine";

#[cfg(not(debug_assertions))]
const PROCESS_PATH_DEFAULT: &str = "./target/release/hash_engine";

pub struct LocalProcess {
    experiment_id: String,
    child: std::process::Child,
    client: Option<nano::Client>,
    engine_url: String,
}

#[async_trait]
impl process::Process for LocalProcess {
    async fn exit_and_cleanup(mut self: Box<Self>) -> Result<()> {
        self.child
            .kill()
            .or_else(|e| match e.kind() {
                std::io::ErrorKind::InvalidInput => Ok(()),
                _ => Err(Report::new(e)),
            })
            .wrap_err("Could not kill the process")?;

        debug!(
            "Cleaned up local engine process for experiment {}",
            &self.experiment_id
        );
        Ok(())
    }

    async fn send(&mut self, msg: &EngineMsg) -> Result<()> {
        // We create the client on the first call here, rather than when the LocalCommand is run,
        // because the engine process needs some time before it's ready to accept NNG connections.
        if self.client.is_none() {
            self.client = Some(nano::Client::new(&self.engine_url, 1)?);
        }
        self.client
            .as_mut()
            .unwrap()
            .send(msg)
            .await
            .map_err(Report::from)
    }
}

pub struct LocalCommand {
    engine_url: String,
    experiment_id: String,
    controller_url: String,
    max_num_workers: usize,
}

impl LocalCommand {
    pub fn new(experiment_id: &str, max_num_workers: usize, controller_url: &str) -> Result<Self> {
        // The NNG URL that the engine process will listen on
        let engine_url = format!("ipc://run-{experiment_id}");

        Ok(LocalCommand {
            engine_url,
            experiment_id: experiment_id.to_string(),
            controller_url: controller_url.to_string(),
            max_num_workers,
        })
    }
}

#[async_trait]
impl process::Command for LocalCommand {
    async fn run(self: Box<Self>) -> Result<Box<dyn process::Process + Send>> {
        let env_process_path = std::env::var("ENGINE_PATH");
        let process_path = if env_process_path.is_err() {
            PROCESS_PATH_DEFAULT
        } else {
            env_process_path.as_ref().unwrap()
        };

        let mut cmd = std::process::Command::new(process_path);
        cmd.arg("--experiment-id")
            .arg(&self.experiment_id)
            .arg("--orchestrator-url")
            .arg(&self.controller_url)
            .arg("--listen-url")
            .arg(&self.engine_url)
            .arg("--max-workers")
            .arg(self.max_num_workers.to_string())
            .stdout(std::process::Stdio::inherit())
            .stderr(std::process::Stdio::inherit());
        debug!("Running `{cmd:?}`");

        let child = cmd
            .spawn()
            .wrap_err_lazy(|| format!("Could not run command: {process_path:?}"))?;
        debug!(
            "Spawned local engine process for experiment {}",
            &self.experiment_id
        );

        Ok(Box::new(LocalProcess {
            experiment_id: self.experiment_id,
            child,
            client: None,
            engine_url: self.engine_url.clone(),
        }))
    }
}
