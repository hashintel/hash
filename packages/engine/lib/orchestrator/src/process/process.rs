use async_trait::async_trait;
use error::Result;
use hash_engine::proto::EngineMsg;

#[async_trait]
pub trait Process {
    async fn exit_and_cleanup(self: Box<Self>) -> Result<()>;
    async fn send(&mut self, msg: &EngineMsg) -> Result<()>;
}

#[async_trait]
pub trait Command {
    async fn run(self: Box<Self>) -> Result<Box<dyn Process + Send>>;
}
