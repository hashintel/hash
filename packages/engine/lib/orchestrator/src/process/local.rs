use std::{
    path::{Path, PathBuf},
    process::ExitStatus,
};

use async_trait::async_trait;
use error::{Report, Result, ResultExt};
use hash_engine_lib::{
    experiment::controller::run::cleanup_experiment,
    proto::{EngineMsg, ExperimentId},
    utils::{LogFormat, LogLevel, OutputLocation},
};

use crate::process;

const ENGINE_BIN_PATH_DEFAULT: &str = env!("CARGO_BIN_FILE_HASH_ENGINE");

#[cfg(debug_assertions)]
const ENGINE_BIN_PATH_FALLBACK: &str = "./target/debug/hash_engine";

#[cfg(not(debug_assertions))]
const ENGINE_BIN_PATH_FALLBACK: &str = "./target/release/hash_engine";

/// A local [`hash_engine`] subprocess using the [`std::process`] library.  
pub struct LocalProcess {
    child: tokio::process::Child,
    client: Option<nano::Client>,
    engine_url: String,
}

#[async_trait]
impl process::Process for LocalProcess {
    async fn exit_and_cleanup(mut self: Box<Self>, experiment_id: ExperimentId) -> Result<()> {
        // Kill the child process as it didn't stop on its own
        let kill_result = self
            .child
            .kill()
            .await
            .or_else(|e| match e.kind() {
                // From `Child::kill` docs: Forces the child process to exit. If the child has
                // already exited, an InvalidInput error is returned
                std::io::ErrorKind::InvalidInput => Ok(()),
                _ => Err(Report::new(e)),
            })
            .wrap_err("Could not kill the process");

        cleanup_experiment(experiment_id);

        debug!("Cleaned up local engine process for experiment");
        kill_result
    }

    /// Creates or reuses a [`nano::Client`] to send a message to the [`hash_engine`] subprocess.
    ///
    /// # Errors
    ///
    /// - if the [`nano::Client`] could not be created
    /// - if the message could not be sent
    async fn send(&mut self, msg: &EngineMsg) -> Result<()> {
        // We create the client on the first call here, rather than when the LocalCommand is run,
        // because the engine process needs some time before it's ready to accept NNG connections.
        if self.client.is_none() {
            self.client = Some(nano::Client::new(&self.engine_url, 1)?);
        }
        self.client
            .as_mut()
            .unwrap()
            .send(msg)
            .await
            .map_err(Report::from)
    }

    async fn wait(&mut self) -> Result<ExitStatus> {
        self.child
            .wait()
            .await
            .wrap_err("Could not wait for the process to exit")
    }
}

/// Stores information to create a [`LocalProcess`] for creating an engine-subprocess.
pub struct LocalCommand {
    experiment_id: ExperimentId,
    engine_url: String,
    controller_url: String,
    num_workers: usize,
    log_format: LogFormat,
    log_level: Option<LogLevel>,
    output_location: OutputLocation,
    log_folder: PathBuf,
    target_max_group_size: Option<usize>,
    js_runner_initial_heap_constraint: Option<usize>,
    js_runner_max_heap_size: Option<usize>,
}

impl LocalCommand {
    #[allow(clippy::too_many_arguments)]
    /// Creates a new [`LocalProcess`] with the provided parameters.
    pub fn new(
        experiment_id: ExperimentId,
        num_workers: usize,
        controller_url: &str,
        log_format: LogFormat,
        log_level: Option<LogLevel>,
        output_location: OutputLocation,
        log_folder: PathBuf,
        target_max_group_size: Option<usize>,
        js_runner_initial_heap_constraint: Option<usize>,
        js_runner_max_heap_size: Option<usize>,
    ) -> Self {
        // The NNG URL that the engine process will listen on
        let engine_url = format!("ipc://run-{experiment_id}");

        Self {
            experiment_id,
            engine_url,
            controller_url: controller_url.to_string(),
            num_workers,
            log_format,
            log_level,
            output_location,
            log_folder,
            target_max_group_size,
            js_runner_initial_heap_constraint,
            js_runner_max_heap_size,
        }
    }
}

#[async_trait]
impl process::Command for LocalCommand {
    /// Spawns an engine process and returns its handle as a [`LocalProcess`].
    ///
    /// # Errors
    ///
    /// - if the process could not be spawned
    async fn run(self: Box<Self>) -> Result<Box<dyn process::Process + Send>> {
        let engine_path = std::env::var("ENGINE_PATH");
        let process_path = if let Ok(process_path) = &engine_path {
            process_path.as_str()
        } else if Path::new(ENGINE_BIN_PATH_DEFAULT).exists() {
            ENGINE_BIN_PATH_DEFAULT
        } else {
            ENGINE_BIN_PATH_FALLBACK
        };

        let mut cmd = tokio::process::Command::new(process_path);
        cmd.arg("--experiment-id")
            .arg(self.experiment_id.to_string())
            .arg("--orchestrator-url")
            .arg(&self.controller_url)
            .arg("--listen-url")
            .arg(&self.engine_url)
            .arg("--num-workers")
            .arg(self.num_workers.to_string())
            .arg("--log-format")
            .arg(self.log_format.to_string())
            .arg("--output")
            .arg(self.output_location.to_string())
            .arg("--log-folder")
            .arg(self.log_folder)
            .stdout(std::process::Stdio::inherit())
            .stderr(std::process::Stdio::inherit());
        if let Some(log_level) = self.log_level {
            cmd.arg("--log-level").arg(log_level.to_string());
        }
        if let Some(target_max_group_size) = self.target_max_group_size {
            cmd.arg("--target-max-group-size")
                .arg(target_max_group_size.to_string());
        }
        if let Some(js_runner_initial_heap_constraint) = self.js_runner_initial_heap_constraint {
            cmd.arg("--js-runner-initial-heap-constraint")
                .arg(js_runner_initial_heap_constraint.to_string());
        }
        if let Some(js_runner_max_heap_size) = self.js_runner_max_heap_size {
            cmd.arg("--js-runner-max-heap-size")
                .arg(js_runner_max_heap_size.to_string());
        }
        debug!("Running `{cmd:?}`");

        let child = cmd
            .spawn()
            .wrap_err_lazy(|| format!("Could not run command: {process_path:?}"))?;
        debug!("Spawned local engine process for experiment");

        Ok(Box::new(LocalProcess {
            child,
            client: None,
            engine_url: self.engine_url.clone(),
        }))
    }
}
