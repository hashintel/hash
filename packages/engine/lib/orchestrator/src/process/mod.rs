//! Functionality to start and communicate with a [`hash_engine`] subprocess.

mod local;

use std::process::ExitStatus;

use async_trait::async_trait;
use error::Result;
use hash_engine_lib::proto::EngineMsg;
use simulation_structure::ExperimentId;

pub use self::local::{LocalCommand, LocalProcess};

/// The engine-subprocess running in the background.
///
/// It's created by a [`Command`] and is used to communicate with the [`hash_engine`] library.
#[async_trait]
pub trait Process {
    /// Exits the subprocess and cleans up resources used for it.
    async fn exit_and_cleanup(self: Box<Self>, experiment_id: ExperimentId) -> Result<()>;

    /// Sends a message to the underlying process
    ///
    /// # Errors
    ///
    /// Returns an error if the message could not be sent.
    async fn send(&mut self, msg: &EngineMsg) -> Result<()>;

    /// Waits for the process to exit completely, returning the status that it exited with.
    async fn wait(&mut self) -> Result<ExitStatus>;
}

/// A command for creating an engine-subprocess represented by [`Process`].
#[async_trait]
pub trait Command {
    /// Creates and runs the subprocess.
    async fn run(self: Box<Self>) -> Result<Box<dyn Process + Send>>;
}
