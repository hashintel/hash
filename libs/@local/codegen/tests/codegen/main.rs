extern crate alloc;

mod enums;
mod lists;
mod maps;
mod structs;
mod tuples;

use std::process::ExitCode;

use hash_codegen::{
    TypeCollection,
    typescript::{TypeScriptGenerator, TypeScriptGeneratorSettings},
};
use insta::assert_snapshot;
use libtest_mimic::{Arguments, Trial};

#[derive(Debug)]
pub enum CodegenTarget {
    Typescript,
}

fn main() -> ExitCode {
    let args = Arguments::from_args();

    let targets = [CodegenTarget::Typescript];

    let mut collection = TypeCollection::default();
    // Register enums
    collection.register::<enums::EnumExternal>();
    collection.register::<enums::EnumInternal>();
    collection.register::<enums::EnumAdjacent>();
    collection.register::<enums::EnumUntagged>();

    // Register structs
    collection.register::<structs::StructUnit>();
    collection.register::<structs::StructUnnamedSingle>();
    collection.register::<structs::StructUnnamedDouble>();
    collection.register::<structs::StructUnnamedTriple>();
    collection.register::<structs::StructSimple>();
    collection.register::<structs::StructEmpty>();
    collection.register::<structs::StructOptional>();
    collection.register::<structs::StructNested>();

    // Register lists
    collection.register::<lists::ListPrimitives>();
    collection.register::<lists::ListOptionals>();
    collection.register::<lists::ListNested>();
    collection.register::<lists::ListStructs>();
    collection.register::<lists::ListCollections>();

    // Register maps
    collection.register::<maps::MapStringKey>();
    collection.register::<maps::MapNumericKey>();
    collection.register::<maps::MapStructValue>();
    collection.register::<maps::MapEnumKey>();
    collection.register::<maps::MapNested>();
    collection.register::<maps::MapOptional>();

    // Register tuples
    collection.register::<tuples::TupleEmpty>();
    collection.register::<tuples::TupleSingle>();
    collection.register::<tuples::TupleDouble>();
    collection.register::<tuples::TupleMultiple>();
    collection.register::<tuples::TupleNested>();
    collection.register::<tuples::TupleOptional>();
    let names = collection
        .iter()
        .map(|(name, _)| name.to_owned())
        .collect::<Vec<_>>();

    let settings = TypeScriptGeneratorSettings::default();
    let generator = TypeScriptGenerator::new(&settings, collection);

    let mut tests = Vec::new();
    for name in &names {
        for target in &targets {
            let test_name = format!("{name}::{target:?}");
            let generated = generator.generate(name);
            tests.push(Trial::test(test_name.clone(), move || {
                assert_snapshot!(test_name, generated);
                Ok(())
            }));
        }
    }

    libtest_mimic::run(&args, tests).exit_code()
}
