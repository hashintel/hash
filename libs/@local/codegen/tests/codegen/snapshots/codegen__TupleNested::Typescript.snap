---
source: libs/@local/codegen/tests/codegen/main.rs
expression: generated
---
export type TupleNested = {
	nested: [string, [number, boolean]]
};
